File: election.py
import time
import sys
import os
import uuid
import csv
from kazoo.client import KazooClient
from kazoo.exceptions import ConnectionLoss, NoNodeError

# --- CONFIGURATION ---
ZK_HOSTS = 'localhost:2181'
ELECTION_PATH = '/election'
DEATH_TIMESTAMP_FILE = 'last_leader_death.txt'
COLD_START_FILE = 'cold_start_data.csv'
FAILOVER_FILE = 'failover_data.csv'

LEADER_TENURE = 15 

class LeaderElectionNode:
    def __init__(self):
        self.zk = KazooClient(hosts=ZK_HOSTS)
        self.zk.start()
        self.my_guid = str(uuid.uuid4())
        self.my_path = None
        self.start_time = time.time()
        self.is_leader = False
        self.running = True
        
    def log_to_csv(self, filename, metric_name, value):
        file_exists = os.path.isfile(filename)
        try:
            with open(filename, 'a', newline='') as f:
                writer = csv.writer(f)
                if not file_exists:
                    writer.writerow(['Metric', 'Seconds'])
                writer.writerow([metric_name, f"{value:.4f}"])
        except Exception as e:
            print(f"[{self.my_guid[:8]}] Error writing to CSV: {e}")

    def volunteer(self):
        try:
            self.zk.ensure_path(ELECTION_PATH)
            self.my_path = self.zk.create(
                f"{ELECTION_PATH}/guid-n_", 
                ephemeral=True, 
                sequence=True
            )
            print(f"[{self.my_guid[:8]}] My Path: {self.my_path}")
            self.check_leadership()
        except ConnectionLoss as e:
            print(f"[{self.my_guid[:8]}] Connection lost during volunteer: {e}")
            sys.exit(1)

    def check_leadership(self):
        if not self.running:
            return
            
        try:
            children = self.zk.get_children(ELECTION_PATH)
        except NoNodeError:
            print(f"[{self.my_guid[:8]}] Election path disappeared, recreating...")
            self.zk.ensure_path(ELECTION_PATH)
            self.volunteer()
            return
        except ConnectionLoss as e:
            print(f"[{self.my_guid[:8]}] Connection lost: {e}")
            return

        if not children:
            print(f"[{self.my_guid[:8]}] No children found, re-volunteering...")
            self.volunteer()
            return

        sorted_children = sorted(children, key=lambda x: int(x.split("_")[-1]))
        my_node_name = self.my_path.split("/")[-1]
        
        if my_node_name not in sorted_children:
            print(f"[{self.my_guid[:8]}] My node disappeared! Re-volunteering...")
            self.volunteer()
            return

        my_index = sorted_children.index(my_node_name)

        # DATA COLLECTION: COLD START
        if len(sorted_children) > 0 and not hasattr(self, 'cold_start_logged'):
            leader_found_time = time.time()
            duration = leader_found_time - self.start_time
            
            if duration < 30:
                print(f"[{self.my_guid[:8]}] Found leader! Cold Start Time: {duration:.4f}s")
                self.log_to_csv(COLD_START_FILE, "Cold_Start_Time", duration)
                self.cold_start_logged = True

        # ELECTION LOGIC
        if my_index == 0:
            self.become_leader()
        else:
            predecessor = sorted_children[my_index - 1]
            predecessor_path = f"{ELECTION_PATH}/{predecessor}"
            print(f"[{self.my_guid[:8]}] I am follower. Watching predecessor: {predecessor}")
            
            if self.zk.exists(predecessor_path, watch=self.predecessor_deleted):
                pass
            else:
                print(f"[{self.my_guid[:8]}] Predecessor already gone!")
                self.check_leadership()

    def predecessor_deleted(self, event):
        print(f"[{self.my_guid[:8]}] Predecessor deleted event received!")
        self.check_leadership()

    def become_leader(self):
        if self.is_leader:
            return  
            
        self.is_leader = True
        print(f"\n[{self.my_guid[:8]}] === I AM THE LEADER ===")
        
        # DATA COLLECTION: FAILOVER
        if os.path.exists(DEATH_TIMESTAMP_FILE):
            try:
                with open(DEATH_TIMESTAMP_FILE, 'r') as f:
                    death_time = float(f.read().strip())
                
                failover_duration = time.time() - death_time
                
                if failover_duration < 60:
                    print(f"[{self.my_guid[:8]}] FAILOVER COMPLETE. Duration: {failover_duration:.4f}s")
                    self.log_to_csv(FAILOVER_FILE, "Failover_Time", failover_duration)
                
                os.remove(DEATH_TIMESTAMP_FILE)
            except Exception as e:
                print(f"[{self.my_guid[:8]}] Error processing failover data: {e}")

        print(f"[{self.my_guid[:8]}] Holding leadership for {LEADER_TENURE} seconds...")
        time.sleep(LEADER_TENURE)
        
        print(f"[{self.my_guid[:8]}] Simulating CRASH now...")
        try:
            with open(DEATH_TIMESTAMP_FILE, 'w') as f:
                f.write(str(time.time()))
        except Exception as e:
            print(f"[{self.my_guid[:8]}] Error writing death timestamp: {e}")
            
        self.running = False
        self.zk.stop()
        os._exit(0)

    def run_as_follower(self):
        try:
            while self.running and not self.is_leader:
                time.sleep(1)
        except KeyboardInterrupt:
            print(f"\n[{self.my_guid[:8]}] Shutting down gracefully...")
            self.cleanup()

    def cleanup(self):
        self.running = False
        try:
            if self.zk:
                self.zk.stop()
                self.zk.close()
        except Exception as e:
            print(f"[{self.my_guid[:8]}] Error during cleanup: {e}")

if __name__ == "__main__":
    node = LeaderElectionNode()
    try:
        node.volunteer()
        
        if not node.is_leader:
            node.run_as_follower()
            
    except KeyboardInterrupt:
        print("\nShutdown requested...")
        node.cleanup()
    except Exception as e:
        print(f"Unexpected error: {e}")
        node.cleanup()
        sys.exit(1)

File: election.go
package main

import (
	//"encoding/csv"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/go-zookeeper/zk"
	"github.com/google/uuid"
)

// --- CONFIGURATION ---
const (
	ZKHosts            = "localhost:2181"
	ElectionPath       = "/election"
	DeathTimestampFile = "last_leader_death.txt"
	ColdStartFile      = "cold_start_data.csv"
	FailoverFile       = "failover_data.csv"
	LeaderTenure       = 15 * time.Second
)

func main() {
	// 1. Connect to Zookeeper
	conn, _, err := zk.Connect([]string{ZKHosts}, time.Second*5)
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	// 2. Create the Election Parent Node (if it doesn't exist)
	exists, _, _ := conn.Exists(ElectionPath)
	if !exists {
		conn.Create(ElectionPath, []byte{}, 0, zk.WorldACL(zk.PermAll))
	}

	// 3. Volunteer (Create my ephemeral node)
	myGUID := uuid.New().String()
	prefix := fmt.Sprintf("%s/guid-n_", ElectionPath)
	myPath, err := conn.Create(prefix, []byte{}, zk.FlagEphemeral|zk.FlagSequence, zk.WorldACL(zk.PermAll))
	if err != nil {
		panic(err)
	}

	fmt.Printf("[%s] My Path: %s\n", myGUID[:8], myPath)
	
	// State variables
	startTime := time.Now()
	coldStartLogged := false
	failoverLogged := false

	// --- MAIN LOOP ---
	// Instead of complex watchers, we just check the status every 100ms.
	for {
		// A. Get all nodes in election
		children, _, err := conn.Children(ElectionPath)
		if err != nil || len(children) == 0 {
			time.Sleep(100 * time.Millisecond)
			continue
		}

		// B. Sort them to find the leader
		sort.Slice(children, func(i, j int) bool {
			return extractSeq(children[i]) < extractSeq(children[j])
		})

		// C. Find where I am in the line
		myIndex := -1
		myNodeName := strings.Split(myPath, "/")[2] // e.g. "guid-n_000000001"
		
		for i, child := range children {
			if child == myNodeName {
				myIndex = i
				break
			}
		}

		if myIndex == -1 {
			// I was deleted? Re-volunteer or exit.
			log.Fatal("My node disappeared!")
		}

		// --- DATA COLLECTION 1: COLD START ---
		if !coldStartLogged {
			duration := time.Since(startTime).Seconds()
			if duration < 30 {
				logToCSV(ColdStartFile, "Cold_Start_Time", duration)
				coldStartLogged = true
			}
		}

		// --- DATA COLLECTION 2: FAILOVER ---
		// Check the death file to see if a leader died recently
		content, err := ioutil.ReadFile(DeathTimestampFile)
		if err == nil {
			tsStr := strings.TrimSpace(string(content))
			if deathTimeFloat, err := strconv.ParseFloat(tsStr, 64); err == nil {
				deathTime := unixFloatToTime(deathTimeFloat)
				failoverDuration := time.Since(deathTime).Seconds()

				// If the crash happened less than 5 seconds ago, and we haven't logged it yet
				if failoverDuration < 5 && !failoverLogged {
					fmt.Printf("[%s] FAILOVER DETECTED. Duration: %.4fs\n", myGUID[:8], failoverDuration)
					logToCSV(FailoverFile, "Failover_Time", failoverDuration)
					failoverLogged = true
				}
			}
		}

		// D. Election Logic
		if myIndex == 0 {
			// I AM THE LEADER
			fmt.Printf("\n[%s] === I AM THE LEADER ===\n", myGUID[:8])
			
			// Stay leader for 15 seconds
			time.Sleep(LeaderTenure)
			
			// Simulate Crash: Write time of death and exit
			writeDeathTimestamp()
			fmt.Println("Simulating CRASH...")
			os.Exit(0) 
		} else {
			// I AM A FOLLOWER
			// Just wait a tiny bit and check again
			time.Sleep(100 * time.Millisecond)
		}
	}
}

// --- HELPER FUNCTIONS ---

func logToCSV(filename, metricName string, value float64) {
	// Simple append to file
	f, err := os.OpenFile(filename, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil { return }
	defer f.Close()

	// Write header only if file is empty
	stat, _ := f.Stat()
	if stat.Size() == 0 {
		f.WriteString("Metric,Seconds\n")
	}

	f.WriteString(fmt.Sprintf("%s,%.4f\n", metricName, value))
}

func writeDeathTimestamp() {
	now := float64(time.Now().UnixNano()) / 1e9
	ioutil.WriteFile(DeathTimestampFile, []byte(fmt.Sprintf("%f", now)), 0644)
}

func unixFloatToTime(val float64) time.Time {
	sec := int64(val)
	nsec := int64((val - float64(sec)) * 1e9)
	return time.Unix(sec, nsec)
}

func extractSeq(nodeName string) int {
	parts := strings.Split(nodeName, "_")
	if len(parts) < 2 { return 0 }
	seq, _ := strconv.Atoi(parts[len(parts)-1])
	return seq
}